hoomd/md/external/wall.py

```diff
@@ -0,0 +1,653 @@
+    if isinstance(wall, hoomd.wall.Sphere):
+        return _md.SphereWall(wall.radius, wall.origin, wall.inside)
+    if isinstance(wall, hoomd.wall.Cylinder):
+        return _md.CylinderWall(wall.radius, wall.origin, wall.axis,
+                                wall.inside)
+    if isinstance(wall, hoomd.wall.Plane):
+        return _md.PlaneWall(wall.origin, wall.normal, wall.inside)
+    raise TypeError(f"Unknown wall type encountered {type(wall)}.")
+
+
+class _WallArrayViewFactory:
+
+    def __init__(self, cpp_wall_potential, wall_type):
+        self.cpp_obj = cpp_wall_potential
+        self.func_name = {
+            hoomd.wall.Sphere: "get_sphere_list",
+            hoomd.wall.Cylinder: "get_cylinder_list",
+            hoomd.wall.Plane: "get_plane_list"
+        }[wall_type]
+
+    def __call__(self):
+        return getattr(self.cpp_obj.field, self.func_name)()
+
+
+class WallPotential(force.Force):
+    r"""Generic wall potential.
+
+    Warning:
+        `WallPotential` should not be used directly.  It is a base class that
+        provides features and documentation common to all standard wall
+        potentials.
+
+    All wall potential commands specify that a given potential energy and
+    potential be computed on all particles in the system within a cutoff
+    distance, :math:`r_{\mathrm{cut}}`, from each wall in the given wall group.
+    The force :math:`\vec{F}` is in the direction of :math:`\vec{r}`, the vector
+    pointing from the particle to the wall or half-space boundary and
+    :math:`V_{\mathrm{pair}}(r)` is the pair potential specified by subclasses
+    of `WallPotential`.  Wall forces are implemented with the concept of
+    half-spaces in mind. There are two modes which are allowed currently in wall
+    potentials: standard and extrapolated.
+
+    .. rubric:: Standard Mode.
+
+    In the standard mode, when :math:`r_{\mathrm{extrap}} \le 0`, the potential
+    energy is only applied to the half-space specified in the wall group.
+    :math:`V(r)` is evaluated in the same manner as when the mode is shift for
+    the analogous :py:mod:`pair <hoomd.md.pair>` potentials within the
+    boundaries of the half-space.
+
+    .. math::
+
+        V(r)  = V_{\mathrm{pair}}(r) - V_{\mathrm{pair}}(r_{\mathrm{cut}})
+
+    For ``inside=True`` (closed) half-spaces:
+
+    .. math::
+        :nowrap:
+
+        \begin{eqnarray*}
+        \vec{F}  = & -\nabla V(r) & 0 \le r < r_{\mathrm{cut}} \\
+                 = & 0 & r \ge r_{\mathrm{cut}} \\
+                 = & 0 & r < 0
+        \end{eqnarray*}
+
+    For ``inside=False`` (open) half-spaces:
+
+    .. math::
+        :nowrap:
+
+        \begin{eqnarray*}
+        \vec{F}  = & -\nabla V(r) & 0 < r < r_{\mathrm{cut}} \\
+                 = & 0 & r \ge r_{\mathrm{cut}} \\
+                 = & 0 & r \le 0
+        \end{eqnarray*}
+
+    .. rubric: Extrapolated Mode:
+
+    The wall potential can be linearly extrapolated beyond a minimum separation
```

Comment:
The figure that explains the extrapolated mode should be shown here instead as it applies to all wall potentials. It would be great to also recreate the figure in the new consistent documentation style.